### PR TITLE
Disegni Image Segmenter PR 2.0

### PR DESCRIPTION
**Pull request by Samu100**
This is the final implementation of my semi-automatic Disegni Segmenter. It has an implementation of the Otsu segmentation method which takes to long to complete, and it also has an implementation of a manual version of the Watershed segmentation method. To use this algorithm, you go on photoshop to find the pixel coordinates of your background and points of interest in your input image. You then hard code these values into main(). Next, you make proper modifications to the command line and working directory. Finally, you run the algorithm and it should spit out your fragmented regions of interest. If you have any questions let me know. :)

### CODE REVIEW

File: apps/src/SegmentDisegni.cpp [hunk 1]
```diff
@@ -61,9 +60,19 @@ int main(int argc, char* argv[])
         return EXIT_FAILURE;
     }
 
+
```
Comment (csparker247) [inline apps/src/SegmentDisegni.cpp:99]: Remove one of these blank lines.

File: apps/src/SegmentDisegni.cpp [hunk 2]
```diff
@@ -61,9 +60,19 @@ int main(int argc, char* argv[])
...
+
     // Run segmenter
     std::cout << "Segmenting image..." << std::endl;
     rt::DisegniSegmenter segmenter;
+
+    //Testing Purpose (Manual Vector Points Given)
+    std::vector<cv::Point> manualPoints = {{181,370}};
+    cv::Point backgroundPoint = {57,55};
```
Comment (csparker247) [inline apps/src/SegmentDisegni.cpp]: This should work.
```suggestion
    std::vector<cv::Point> manualPoints{{181,370}};
    cv::Point backgroundPoint{57,55};
```

File: rtlib/include/rt/DisegniSegmenter.hpp [hunk 3]
```diff
@@ -29,6 +29,14 @@ class DisegniSegmenter
     void setPreprocessSharpen(bool b);
     /** @brief Median blur input image prior to segmentation */
     void setPreprocessBlur(bool b);
+    //Set for Contours
+    void setContours(std::vector<cv::Point> b);
```
Comment (csparker247) [inline rtlib/include/rt/DisegniSegmenter.hpp]:
```suggestion
    void setForegroundCoords(std::vector<cv::Point> b);
```

File: rtlib/include/rt/DisegniSegmenter.hpp [hunk 4]
```diff
@@ -29,6 +29,14 @@ class DisegniSegmenter
     void setPreprocessSharpen(bool b);
     /** @brief Median blur input image prior to segmentation */
     void setPreprocessBlur(bool b);
+    //Set for Contours
```
Comment (csparker247) [inline rtlib/include/rt/DisegniSegmenter.hpp]: Use doxygen-style comments for the new members in the header.

File: rtlib/include/rt/DisegniSegmenter.hpp [hunk 5]
```diff
@@ -51,6 +59,14 @@ class DisegniSegmenter
     cv::Mat labeled_;
     /** Segmented subregions */
     std::vector<cv::Mat> results_;
+    //Contours boundary vector
+    std::vector<cv::Point> contours_;
```
Comment (csparker247) [inline rtlib/include/rt/DisegniSegmenter.hpp]:
```suggestion
    std::vector<cv::Point> fgCoords_;
```

File: rtlib/include/rt/DisegniSegmenter.hpp [hunk 6]
```diff
@@ -63,7 +79,10 @@ class DisegniSegmenter
     cv::Mat preprocess_();
 
     /** Run watershed on image */
-    static cv::Mat watershed_image_(const cv::Mat& input);
+    cv::Mat watershed_image_(const cv::Mat& input);
+
+    //Otsu Image Segmentation
+    cv::Mat otsu_segmentation_(const cv::Mat& input);
```
Comment (csparker247) [inline rtlib/include/rt/DisegniSegmenter.hpp]: Remove the Otsu stuff from this branch. We'll add it from the other branch if we keep it.

File: rtlib/src/DisegniSegmenter.cpp [hunk 7]
```diff
@@ -4,15 +4,27 @@
 #include <limits>
 #include <map>
 #include <set>
+#include <iostream>
 
 #include <opencv2/imgproc.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
```
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]:
```suggestion
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
```

File: rtlib/src/DisegniSegmenter.cpp [hunk 8]
```diff
@@ -156,53 +142,89 @@ cv::Mat DisegniSegmenter::watershed_image_(const cv::Mat& input)
...
+            auto width = i.second.br.x - i.second.tl.x;
+            std::cout << idx++ << ": " << i.second.tl.x << " " << i.second.tl.y << " " << width << " " << height << std::endl;
+            cv::Rect roi(i.second.tl.x, i.second.tl.y, width, height);
+            cv::Mat subimg = input(roi).clone();
+            subimgs.push_back(subimg);
+        }
+        std::cout << std::endl;
 
-    return subimgs;
+        return subimgs;
```
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]: Revert all of the changes to this code block.

File: rtlib/src/DisegniSegmenter.cpp [hunk 9]
```diff
@@ -4,15 +4,27 @@
...
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <itkOpenCVImageBridge.h>
+#include <rt/ImageTypes.hpp>
+#include <itkScalarImageToHistogramGenerator.h>
+#include <itkBinaryThresholdImageFilter.h>
+#include <itkOtsuMultipleThresholdsCalculator.h>
+#include <itkOtsuMultipleThresholdsImageFilter.h>
+#include <itkRescaleIntensityImageFilter.h>
+#include <highgui.h>
```
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]:
```suggestion
```

File: rtlib/src/DisegniSegmenter.cpp [hunk 10]
```diff
@@ -4,15 +4,27 @@
 #include <limits>
 #include <map>
 #include <set>
+#include <iostream>
```
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]:
```suggestion
```

File: rtlib/src/DisegniSegmenter.cpp (hunk 9 above)
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]: Review the [Header Includes](https://github.com/viscenter/style-guides/blob/master/C%2B%2B%20Style%20Guide.md#header-includes) section of the style guide and update accordingly.

File: rtlib/src/DisegniSegmenter.cpp [hunk 11]
```diff
@@ -109,45 +123,17 @@ cv::Mat DisegniSegmenter::preprocess_()
...
-    for (size_t i = 0; i < contours.size(); i++) {
-        cv::drawContours(
-            markers, contours, static_cast<int>(i),
-            cv::Scalar(static_cast<int>(i) + 1), -1);
+    cv::Mat markers = cv::Mat::zeros(input.size(), CV_32S);
+
+    // Draw the foreground markers with circles
+    for (size_t i = 0; i < contours_.size(); i++) {
+        cv::circle(markers, contours_[i], 1, cv::Scalar(static_cast<int>(i) + 1), -1);
     }
```
Comment (csparker247) [inline rtlib/src/DisegniSegmenter.cpp]:
```suggestion
    int label = 1;
    for (const auto& coord : fgCoords_) {
        cv::circle(markers, coord, 1, cv::Scalar(label++), -1);
    }
```

**Review by csparker247 — CHANGES_REQUESTED** [review]

File: rtlib/src/DisegniSegmenter.cpp (hunk 10 above)
Comment (Samu100) [reply rtlib/src/DisegniSegmenter.cpp]: Did you mean to leave this comment?

Comment (csparker247) [reply rtlib/src/DisegniSegmenter.cpp]: Yes. You shouldn't need iostream once you remove the `std::cout` stuff

Comment (csparker247) [reply rtlib/src/DisegniSegmenter.cpp]: Oh, sorry. No. I meant to remove the one in your class, not here.

Comment (csparker247) [reply rtlib/src/DisegniSegmenter.cpp]: I don't know what's happening. GitHub keeps showing this under different files. Remove the iostream include and all uses of `std::cout` from your class.

**Review by csparker247 — APPROVED** [review]

**Comment by csparker247** [comment]
Final implementations adds a number of improvements over the original approach. Rather than trying to automatically seed the watershed algorithm, `DisegniSegmenter` now has an interface for defining manual seed points for both the background and foreground objects. There are additional interface improvements for controlling seed point size and for adding buffer space to segmented subregions. The utility `rt_segment_disegni` has also been updated to support all of these features and now enables saving the raw label image to a 32-bit, signed integer TIFF image.

**Pull request merged**